drivers/net/can/sja1000/ts7800v2_can.c

```diff
@@ -0,0 +1,287 @@
+/*
+ * ts7800v2_can.c - Interface for the CAN controller in the TS-7800v2's FPGA.
+ *
+ * Copyright (c) 2018, Technologic Systems, Inc.
+ * All rights reserved.
+ *
+ *  LICENCE:
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation version 2.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/netdevice.h>
+#include <linux/of.h>
+#include <linux/of_irq.h>
+#include <linux/io.h>
+#include <linux/platform_device.h>
+#include <linux/if_arp.h>
+#include <linux/if_ether.h>
+#include <linux/can.h>
+#include <linux/can/dev.h>
+#include <linux/can/platform/sja1000.h>
+#include <linux/can/skb.h>
+#include <linux/slab.h>
+#include <net/rtnetlink.h>
+
+#include "sja1000.h"
+
+#define DEBUG_ON 0
+
+static DEFINE_SPINLOCK(fpga_lock);
+
+/**
+   There is a 32-bit register in the syscon at offset 0x4c.
+
+   CAN has its own 256 byte 8-bit bus address space manipulated via this
+   one register.
+
+   Bit 31: Start CAN bus cycle
+   Bit 30: CAN bus cycle is a write
+   Bits 15-8: CAN data for read or write
+   Bits 7-0: CAN address
+
+   To initiate a CAN read for (e.g.) address 0x55, first write syscon
+   register 0x4c with 0x800000aa.  Then read back the same register.  When
+   bit 31 is clear, the bus cycle is complete and the data is in bits 15-8.
+
+   To initiate a CAN write of value 0xaa for address 0x55, write
+   0xc000aa55.
+
+*/
+
+#define DRIVER_NAME	"ts7800v2_can"
+#define TX_MAX 1
+#define RX_MAX 1
+#define CAN_CNTL_REG_OFFSET 0x4C
+#define CAN_CNTL_START BIT(31)
+#define CAN_CNTL_WRITE BIT(30)
+
+
+static u8 ts7800v2_read_reg(const struct sja1000_priv *priv, int reg)
+{
+   unsigned long flags;
+   volatile unsigned int *syscon = (unsigned int *)priv->reg_base;
+   unsigned int v;
+
+   spin_lock_irqsave(&fpga_lock, flags);
+
+   while(readl(&syscon[CAN_CNTL_REG_OFFSET / 4]) & CAN_CNTL_START);
+
+   writel(CAN_CNTL_START | reg, &syscon[CAN_CNTL_REG_OFFSET / 4]);
+
+   do {
+      v = readl(&syscon[CAN_CNTL_REG_OFFSET / 4]);
+   } while(v & CAN_CNTL_START);
+
+#if (DEBUG_ON)
```

Comment:
I'd assume we want all instances of `DEBUG_ON` to follow kernel paradigm of just `DEBUG` but I'm not sure where `DEBUG_ON` originated.

Reply:
Maybe @markfeathers can comment on this?